bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java

```diff
@@ -1231,58 +1233,33 @@ public static Type resolveMatchingTypeForUnion(Object value, Type type) {
             return PredefinedTypes.TYPE_BOOLEAN;
         }
 
-        if (checkIsLikeType(value, PredefinedTypes.TYPE_BYTE)) {
-            return PredefinedTypes.TYPE_BYTE;
-        }
-
         if (checkIsLikeType(value, PredefinedTypes.TYPE_DECIMAL)) {
             return PredefinedTypes.TYPE_DECIMAL;
         }
 
-        Type anydataArrayType = new BArrayType(type, type.isReadOnly());
-        if (checkIsLikeType(value, anydataArrayType)) {
-            return anydataArrayType;
-        }
+        boolean readOnlyTargetType = targetUnionType.isReadOnly();
 
         if (checkIsLikeType(value, PredefinedTypes.TYPE_XML)) {
-            return PredefinedTypes.TYPE_XML;
+            return readOnlyTargetType ? PredefinedTypes.TYPE_READONLY_XML : PredefinedTypes.TYPE_XML;
         }
 
-        Type anydataMapType = new BMapType(type, type.isReadOnly());
-        if (checkIsLikeType(value, anydataMapType)) {
-            return anydataMapType;
+        Type targetTypeArrayType = new BArrayType(targetUnionType, readOnlyTargetType);
+        if (checkIsLikeType(value, targetTypeArrayType)) {
+            return targetTypeArrayType;
         }
 
-        //not possible
-        return null;
-    }
-
-    private static boolean isDeepConversionRequiredForArray(Type sourceType) {
-        Type elementType = TypeUtils.getReferredType(((BArrayType) sourceType).getElementType());
-
-        if (elementType != null) {
-            if (TypeUtils.isValueType(elementType)) {
-                return false;
-            } else if (elementType instanceof BArrayType) {
-                return isDeepConversionRequiredForArray(elementType);
-            }
-            return true;
+        Type targetTypeMapType = new BMapType(targetUnionType, readOnlyTargetType);
+        if (checkIsLikeType(value, targetTypeMapType)) {
+            return targetTypeMapType;
         }
-        return true;
-    }
-
-    private static boolean isDeepConversionRequiredForMap(Type sourceType) {
```

Comment:
Why we remove these methods?

Reply:
Places which call these methods are redundant. I have removed those. Hence these are not used anymore. 
Eg:
```java
if (value instanceof MapValue && ((MapValue) value).getType().getTag() == TypeTags.MAP_TAG &&
        !isDeepConversionRequiredForMap(((MapValue) value).getType())) {
    return TypeCreator.createMapType(type); // returns map<anydata> when target type is anydata
}
```
The above thing will be now handed in 
```java
Type targetTypeMapType = new BMapType(targetUnionType, readOnlyTargetType);
if (checkIsLikeType(value, targetTypeMapType)) {
    return targetTypeMapType; // returns map<anydata> when target type is anydata
}
```